File: Patterns/Composite Pattern/Cpp/Menu/Menu.cpp
#pragma once

#include <iostream>
#include <string>
#include <list>
#include "MenuComponent.cpp"

using namespace std;

class Menu : public MenuComponent {
private:
    list<MenuComponent*> menuComponents;
    string name;
    string description;

public:
    Menu(string name, string description) : name(name), description(description) {
        menuComponents = list<MenuComponent*>();
    }

    void add(MenuComponent* menuComponent) override {
        menuComponents.push_back(menuComponent);
    }

    void remove(MenuComponent* menuComponent) override {
        menuComponents.remove(menuComponent);
    }

    MenuComponent* getChild(int i) override {
        auto iter = menuComponents.begin();
        advance(iter, i);
        return *iter;
    }

    string getName() override {
        return name;
    }

    string getDescription() override {
        return description;
    }

    void print() {
        cout << "\n" << getName();
        cout << ", " << getDescription() << "\n";
        cout << "----------------------" << "\n";

        for (auto& menuComponent: menuComponents) {
            menuComponent->print();
        }
    }
};

File: Patterns/Composite Pattern/CSharp/Menu/Menu.cs
using System;
using System.Collections.Generic;

namespace Menu
{
    public class Menu : MenuComponent
    {
        private List<MenuComponent> menuComponents;
        private string? name;
        private string? description;

        public Menu(string? name, string? description) : base() {
            menuComponents = new List<MenuComponent>();
            Name = name;
            Description = description;
        }

        public override void Add(MenuComponent menuComponent) => menuComponents.Add(menuComponent);

        public override void Remove(MenuComponent menuComponent) => menuComponents.Remove(menuComponent);

        public override MenuComponent? GetChild(int i) => menuComponents[i];

        public override string? Name { get => name ?? "NoText"; set => name = value; }

        public override string? Description { get => description ?? "NoText"; set => description = value; }

        public override void Print() {
            Console.WriteLine($"\n{name}, {description}");
            Console.WriteLine("----------------------");

            foreach (var menuComponent in menuComponents) {
                menuComponent.Print();
            }
        }
    }
}